File: bestia-zoneserver/src/test/java/net/bestia/zoneserver/entity/EntityServiceTest.java
package net.bestia.zoneserver.entity;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import com.hazelcast.core.HazelcastInstance;

import net.bestia.zoneserver.BasicMocks;
import net.bestia.zoneserver.entity.components.PositionComponent;
import net.bestia.zoneserver.entity.components.VisibleComponent;

public class EntityServiceTest {

	private EntityService entityService;

	private BasicMocks basicMocks = new BasicMocks();
	private HazelcastInstance hz = basicMocks.hazelcastMock();

	@Before
	public void setup() {
		entityService = new EntityService(hz);
	}

	@Test
	public void newEntity_returnsNewEntity() {
		Entity e1 = entityService.newEntity();
		Entity e2 = entityService.newEntity();
		Assert.assertNotEquals(0, e1.getId());
		Assert.assertNotEquals(0, e2.getId());
		Assert.assertNotEquals(e1, e2);
		Assert.assertNotNull(e1);
	}

	@Test(expected = NullPointerException.class)
	public void addComponent_nullComp_throws() {
		Entity e1 = entityService.newEntity();
		entityService.addComponent(e1, null);
	}

	@Test
	public void addComponent_entityRefAndComponent_isAdded() {
		Entity e1 = entityService.newEntity();
		PositionComponent posComp = entityService.addComponent(e1, PositionComponent.class);
		Assert.assertNotNull(posComp);
		Assert.assertTrue(entityService.hasComponent(e1, PositionComponent.class));
	}

	@Test(expected = NullPointerException.class)
	public void delete_null_throws() {
		entityService.delete(null);
	}

	@Test
	public void delete_notSavedEntity_doesNothing() {
		Entity e = entityService.newEntity();
		entityService.delete(e);
	}

	@Test
	public void delete_savedEntity_deletesIt() {
		Entity e = entityService.newEntity();
		entityService.delete(e);
		Entity e2 = entityService.getEntity(e.getId());
		Assert.assertNull(e2);
	}

	@Test
	public void save_entity_savedIt() {
		Entity e = entityService.newEntity();
		Entity e2 = entityService.getEntity(e.getId());
		Assert.assertEquals(e, e2);
	}

	@Test
	public void addComponent_entityIdAndComponent_isAdded() {
		Entity e1 = entityService.newEntity();
		PositionComponent posComp = entityService.addComponent(e1, PositionComponent.class);
		Assert.assertNotNull(posComp);
		Assert.assertTrue(entityService.hasComponent(e1, PositionComponent.class));
	}

	@Test
	public void hasComponent_notAddedComponent_false() {
		Entity e1 = entityService.newEntity();
		Assert.assertFalse(entityService.hasComponent(e1, VisibleComponent.class));
	}

	@Test
	public void hasComponent_addedComponent_true() {
		Entity e1 = entityService.newEntity();
		entityService.addComponent(e1, PositionComponent.class);
		Assert.assertTrue(entityService.hasComponent(e1, PositionComponent.class));
	}
}

File: bestia-zoneserver/src/test/kotlin/net/bestia/zoneserver/entity/EntityServiceTest.kt
package net.bestia.zoneserver.entity

import akka.actor.PoisonPill
import com.nhaarman.mockito_kotlin.check
import com.nhaarman.mockito_kotlin.verify
import net.bestia.messages.entity.EntityEnvelope
import net.bestia.model.bestia.Direction
import net.bestia.zoneserver.MessageApi
import net.bestia.zoneserver.entity.component.PositionComponent
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.junit.MockitoJUnitRunner
import net.bestia.model.geometry.Point

@RunWith(MockitoJUnitRunner::class)
class EntityServiceTest {

  @Mock
  private lateinit var messageApi: MessageApi

  private lateinit var entityService: EntityService

  @Before
  fun setup() {
    entityService = EntityService(IdGeneratorService(), messageApi)
  }

  @Test
  fun `newEntity generated new entity with id not 0`() {
    val e = entityService.newEntity()
    Assert.assertNotEquals(0, e.id)
  }

  @Test
  fun `delete() sends PoisonPill to entity`() {
    val e = entityService.newEntity()
    entityService.delete(e)

    verify(messageApi).send(check {
      Assert.assertTrue(it is EntityEnvelope)
      it as EntityEnvelope
      Assert.assertEquals(e.id, it.entityId)
      Assert.assertTrue(it.content is PoisonPill)
    })
  }

  @Test
  fun `updateComponent send entity envelope with component`() {
    val posComp = PositionComponent(
        1337,
        isSightBlocking = false,
        facing = Direction.NORTH,
        shape = Point(10, 10)
    )

    entityService.updateComponent(posComp)

    verify(messageApi).send(check {
      Assert.assertTrue(it is EntityEnvelope)
      it as EntityEnvelope
      Assert.assertEquals(1337, it.entityId)
      Assert.assertEquals(posComp, it.content)
    })
  }
}